### ISSUE TITLE
`estimateEffect()` gives different results for single topic vs. multiple topics

### ISSUE THREAD

**Issue by jooyoungseo** [bstewart/stm#254]
In the following reproducible example, shouldn't the estimated results be the same in theory? Because each regression is run separately. However, the result is different. Is it a bug? Or, did I understand incorrectly?

``` r
library(stm)
#> stm v1.3.6 successfully loaded. See ?stm for help. 
#>  Papers, resources, and other materials at structuraltopicmodel.com
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(tidytext)

# Case 1: Estimate only one topic
prep1 <- estimateEffect(c(1) ~ treatment, gadarianFit, gadarian)

# Tidy regression table
td_prep1 <- tidytext::tidy(prep1)

# Print
td_prep1
#> # A tibble: 2 x 6
#>   topic term        estimate std.error statistic  p.value
#>   <int> <chr>          <dbl>     <dbl>     <dbl>    <dbl>
#> 1     1 (Intercept)    0.444    0.0227     19.5  2.17e-57
#> 2     1 treatment     -0.161    0.0300     -5.36 1.53e- 7

# Case 2: Estimate three topics
prep2 <- estimateEffect(1:3 ~ treatment, gadarianFit, gadarian)

# Tidy regression table
td_prep2 <- tidytext::tidy(prep2) %>%
  # Filter topic 1
  filter(topic == 1)

# Print
td_prep2
#> # A tibble: 2 x 6
#>   topic term        estimate std.error statistic  p.value
#>   <int> <chr>          <dbl>     <dbl>     <dbl>    <dbl>
#> 1     1 (Intercept)    0.444    0.0229     19.3  1.09e-56
#> 2     1 treatment     -0.162    0.0303     -5.35 1.65e- 7

# Compare
identical(td_prep1, td_prep2)
#> [1] FALSE
```

<sup>Created on 2021-05-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> - Session info ---------------------------------------------------------------
#>  setting  value                       
#>  version  R version 4.0.5 (2021-03-31)
#>  os       Windows 10 x64              
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  English_United States.1252  
#>  ctype    English_United States.1252  
#>  tz       America/New_York            
#>  date     2021-05-14                  
#> 
#> - Packages -------------------------------------------------------------------
#>  package     * version date       lib source                       
#>  assertthat    0.2.1   2019-03-21 [1] CRAN (R 4.0.4)               
#>  backports     1.2.1   2020-12-09 [1] CRAN (R 4.0.3)               
#>  cli           2.5.0   2021-04-26 [1] CRAN (R 4.0.5)               
#>  crayon        1.4.1   2021-02-08 [1] CRAN (R 4.0.4)               
#>  data.table    1.14.0  2021-02-21 [1] CRAN (R 4.0.4)               
#>  DBI           1.1.1   2021-01-15 [1] CRAN (R 4.0.4)               
#>  digest        0.6.27  2020-10-24 [1] CRAN (R 4.0.4)               
#>  dplyr       * 1.0.6   2021-05-05 [1] CRAN (R 4.0.5)               
#>  ellipsis      0.3.2   2021-04-29 [1] CRAN (R 4.0.5)               
#>  evaluate      0.14    2019-05-28 [1] CRAN (R 4.0.4)               
#>  fansi         0.4.2   2021-01-15 [1] CRAN (R 4.0.4)               
#>  fs            1.5.0   2020-07-31 [1] CRAN (R 4.0.4)               
#>  generics      0.1.0   2020-10-31 [1] CRAN (R 4.0.4)               
#>  glue          1.4.2   2020-08-27 [1] CRAN (R 4.0.4)               
#>  highr         0.9     2021-04-16 [1] CRAN (R 4.0.5)               
#>  htmltools     0.5.1.1 2021-01-22 [1] CRAN (R 4.0.5)               
#>  janeaustenr   0.1.5   2017-06-10 [1] CRAN (R 4.0.4)               
#>  knitr         1.33    2021-04-24 [1] CRAN (R 4.0.5)               
#>  lattice       0.20-44 2021-05-02 [1] CRAN (R 4.0.5)               
#>  lifecycle     1.0.0   2021-02-15 [1] CRAN (R 4.0.4)               
#>  magrittr      2.0.1   2020-11-17 [1] CRAN (R 4.0.4)               
#>  Matrix        1.2-18  2019-11-27 [1] CRAN (R 4.0.3)               
#>  matrixStats   0.58.0  2021-01-29 [1] CRAN (R 4.0.4)               
#>  pillar        1.6.0   2021-04-13 [1] CRAN (R 4.0.5)               
#>  pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.0.4)               
#>  ps            1.6.0   2021-02-28 [1] CRAN (R 4.0.4)               
#>  purrr         0.3.4   2020-04-17 [1] CRAN (R 4.0.4)               
#>  R6            2.5.0   2020-10-28 [1] CRAN (R 4.0.4)               
#>  Rcpp          1.0.6   2021-01-15 [1] CRAN (R 4.0.4)               
#>  reprex        2.0.0   2021-04-02 [1] CRAN (R 4.0.5)               
#>  rlang         0.4.11  2021-04-30 [1] CRAN (R 4.0.5)               
#>  rmarkdown     2.8     2021-05-07 [1] CRAN (R 4.0.5)               
#>  rstudioapi    0.13    2020-11-12 [1] CRAN (R 4.0.4)               
#>  sessioninfo   1.1.1   2018-11-05 [1] CRAN (R 4.0.4)               
#>  SnowballC     0.7.0   2020-04-01 [1] CRAN (R 4.0.3)               
#>  stm         * 1.3.6   2021-05-10 [1] Github (bstewart/stm@7750a91)
#>  stringi       1.5.3   2020-09-09 [1] CRAN (R 4.0.3)               
#>  stringr       1.4.0   2019-02-10 [1] CRAN (R 4.0.4)               
#>  styler        1.4.1   2021-03-30 [1] CRAN (R 4.0.4)               
#>  tibble        3.1.1   2021-04-18 [1] CRAN (R 4.0.5)               
#>  tidyselect    1.1.1   2021-04-30 [1] CRAN (R 4.0.5)               
#>  tidytext    * 0.3.1   2021-04-10 [1] CRAN (R 4.0.5)               
#>  tokenizers    0.2.1   2018-03-29 [1] CRAN (R 4.0.4)               
#>  utf8          1.2.1   2021-03-12 [1] CRAN (R 4.0.4)               
#>  vctrs         0.3.8   2021-04-29 [1] CRAN (R 4.0.5)               
#>  withr         2.4.2   2021-04-18 [1] CRAN (R 4.0.5)               
#>  xfun          0.22    2021-03-11 [1] CRAN (R 4.0.4)               
#>  yaml          2.2.1   2020-02-01 [1] CRAN (R 4.0.4)               
#> 
#> [1] C:/Program Files/R/R-4.0.5/library
```

</details>

**Comment by brhkim** [comment]
This might be too late to be useful, but are you setting a consistent random seed in-between those two runs (e.g., `set.seed(1234)`)? 

My understanding is that the estimateEffect() function randomly samples topic proportions from the estimated topic proportion distributions for each document over and over again to estimate any given effect in a Bayesian manner: "Mechanically this means we draw a set of topic proportions from the variational posterior, compute our coefficients, then repeat. To compute quantities of interest we simulate within each batch of coefficients and then average over all our results" (https://search.r-project.org/CRAN/refmans/stm/html/estimateEffect.html).

In other words, the estimation process is stochastic and will change due to this randomness in the sampling process. So we wouldn't expect *any* two separate runs of estimateEffect() to produce exactly the same results. You can either try setting the seed between runs to see if they then produce the same results, or just run the exact same call twice in a row and observe that they won't be the same.
